java/ql/src/experimental/Security/CWE/CWE-1004/SensitiveCookieNotHttpOnly.ql

```diff
@@ -0,0 +1,191 @@
+/**
+ * @name Sensitive cookies without the HttpOnly response header set
+ * @description Sensitive cookies without 'HttpOnly' leaves session cookies vulnerable to an XSS attack.
+ * @kind path-problem
+ * @id java/sensitive-cookie-not-httponly
+ * @tags security
+ *       external/cwe/cwe-1004
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSteps
+import semmle.code.java.frameworks.Servlets
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/** Gets a regular expression for matching common names of sensitive cookies. */
+string getSensitiveCookieNameRegex() { result = "(?i).*(auth|session|token|key|credential).*" }
+
+/** Holds if a string is concatenated with the name of a sensitive cookie. */
+predicate isSensitiveCookieNameExpr(Expr expr) {
+  expr.(CompileTimeConstantExpr)
+      .getStringValue()
+      .toLowerCase()
+      .regexpMatch(getSensitiveCookieNameRegex()) or
+  isSensitiveCookieNameExpr(expr.(AddExpr).getAnOperand())
+}
+
+/** Holds if a string is concatenated with the `HttpOnly` flag. */
+predicate hasHttpOnlyExpr(Expr expr) {
+  expr.(CompileTimeConstantExpr).getStringValue().toLowerCase().matches("%httponly%") or
+  hasHttpOnlyExpr(expr.(AddExpr).getAnOperand())
+}
+
+/** The method call `Set-Cookie` of `addHeader` or `setHeader`. */
+class SetCookieMethodAccess extends MethodAccess {
+  SetCookieMethodAccess() {
+    (
+      this.getMethod() instanceof ResponseAddHeaderMethod or
+      this.getMethod() instanceof ResponseSetHeaderMethod
+    ) and
+    this.getArgument(0).(CompileTimeConstantExpr).getStringValue().toLowerCase() = "set-cookie"
+  }
+}
+
+/** The cookie class of Java EE. */
+class CookieClass extends RefType {
+  CookieClass() {
+    this.getASupertype*()
+        .hasQualifiedName(["javax.servlet.http", "javax.ws.rs.core", "jakarta.ws.rs.core"], "Cookie")
+  }
+}
+
+/** The method call `toString` to get a stringified cookie representation. */
+class CookieInstanceExpr extends TaintPreservingCallable {
+  CookieInstanceExpr() {
+    this.getDeclaringType() instanceof CookieClass and
+    this.hasName("toString")
+  }
+
+  override predicate returnsTaintFrom(int arg) { arg = -1 }
+}
+
+/** Sensitive cookie name used in a `Cookie` constructor or a `Set-Cookie` call. */
+class SensitiveCookieNameExpr extends Expr {
+  SensitiveCookieNameExpr() {
+    exists(
+      ClassInstanceExpr cie // new Cookie("jwt_token", token)
+    |
+      cie.getConstructedType() instanceof CookieClass and
+      this = cie and
+      isSensitiveCookieNameExpr(cie.getArgument(0))
+    )
+    or
+    exists(
+      SetCookieMethodAccess ma // response.addHeader("Set-Cookie: token=" +authId + ";HttpOnly;Secure")
+    |
+      this = ma.getArgument(1) and
+      isSensitiveCookieNameExpr(this)
+    )
+  }
+}
```

Comment:
I think the source should just be any `Expr` that satisfies `isSensitiveCookieNameExpr`.

The problem with doing it the way you have here is that it if a sensitive cookie name is assigned to a variable, and then the variable is used in one of these two situations, then I don't think it will be caught. You should make use of the power of the taint tracking library to follow expressions as they flow around (even between functions).

For the first case here, that means you'll have to tell the taint tracking library that a constructor for a `CookieClass` returns a tainted instance if its argument 0 is tainted. There is something very similar for `java.lang.String`, so you can see what I mean: https://github.com/github/codeql/blob/8606386c2ce297221e08ff9cf81c261225a0d1c5/java/ql/src/semmle/code/java/dataflow/FlowSteps.qll#L75 .
For the second case you won't need any extra logic, as argument 1 of `SetCookieMethodAccess` is already a sink.

Reply:
Thanks @owen-mc for the detailed guidance. I've made the change.